### ISSUE TITLE
[BUG] Cannot backup remote dir given as symlink in the path /dir1/dir2/symlink -> /dir1/dir2/dir3/dir4

### ISSUE THREAD

**Issue by andreas-pgss** [rdiff-backup/rdiff-backup#787]
## Bug summary

I cannot backup a remote directory because it is & needs to bei given as symlink:

> /dir1/dir2/symlink -> /dir1/dir2/dir3/dir4

If I try:

```bash
rdiff-backup user@remote.host::/dir1/dir2/symlink localdir
```

I get:

```bash
Fatal Error: Source /dir1/dir2/symlink is not a directory
```

Using the "real path" works

```
rdiff-backup user@remote.host::/dir1/dir2/dir3/dir4 localdir
```

but I cannot use this as it would break the whole logic of the script I am trying to use

## Version, Python, Operating System

Call `rdiff-backup info` (version >= 2.1) or `rdiff-backup -v9` and replace the following line with the output, repeat for each environment impacted:

```yaml
Local host (destination):
2022-11-18 14:05:14.158226 +0100  <CLIENT-6584>  Using rdiff-backup version 2.0.5
2022-11-18 14:05:14.158324 +0100  <CLIENT-6584>         with cpython /usr/bin/python3 version 3.8.10
2022-11-18 14:05:14.173891 +0100  <CLIENT-6584>         on Linux-5.4.0-132-generic-x86_64-with-glibc2.29, fs encoding utf-8

Remote host (source):
2022-11-18 14:06:27.027946 +0100  <CLIENT-11628>  Using rdiff-backup version 2.0.5
2022-11-18 14:06:27.028050 +0100  <CLIENT-11628>        with cpython /usr/bin/python3 version 3.5.3
2022-11-18 14:06:27.032400 +0100  <CLIENT-11628>        on Linux-4.9.0-17-amd64-x86_64-with-debian-9.13, fs encoding utf-8
```

## rdiff-backup call

```bash
rdiff-backup user@remote.host::/dir1/dir2/symlink localdir
```

## What happened and what did you expect?

I got

```bash
Fatal Error: Source /dir1/dir2/symlink is not a directory
```

when I expected this to work, as the target of the symlink is a directory and rsync transfers this just fine.

It would be great if rdiff-backup would handle this just like rsync (whose librsync it is using). I think if the path of source or target directory ends in a symlink, rdiff-backup should just check if the symlink points to a directory and only complain if the symlink's target is not a directory.

**Comment by ericzolf** [comment]
I'd say, works as designed, but we can consider this as an enhancement. The issue is that there are IMHO multiple places expecting a directory, and this will be a special case under Windows.

Two questions:
1. What are you trying to reach?
2. did you try with a slash at the end?

**Comment by andreas-pgss** [comment]
Hi Eric,

> I'd say, works as designed, but we can consider this as an enhancement. The issue is that there are IMHO multiple places expecting a directory, and this will be a special case under Windows.

I am using this on Linux machines (on both ends) where symlinks are very common. In fact, I am trying to set up backups of vhosts on a webserver using ISPConfig for management of the websites, so the path for each website is /var/www/<fqdn>, the "fqdn" part being a symlink to /var/www/clients/client<m>/web<n>.

I do not see how this is working "as expected" as even the tool you are using internally via librsync works fine with paths ending in a symlink to a directory. To me it looks like a check that is basically a good thing (the source should be a directory) falls short of considering a quite common case. That was why I considered it a bug.

> Two questions:
> 
>     1. What are you trying to reach?

I want to have reverse differential backups of the vhosts' web directories and can only pull from the local machine as there is no way the external webserver can (or should be able to) reach the local, internal machine.

>     2. did you try with a slash at the end?

Yes, I did, but this did not change the behaviour.
I also tried to backup from a path like /dir1/dir2/symlink/dir3 to local storage. That does work.
But it is not feasible with the websites as managed by ISPConfig. So, for now, I have no way of using rdiff-backup for this purpose, simply because of the symlinks at the end of the paths I can use.

**Comment by ericzolf** [comment]
"Works as designed" doesn't mean that the design was good, just that the program works as it was designed; and it doesn't mean either that it works as expected _by users_, only that it works as expected by someone who knows the design resp. the code. It also doesn't mean that it won't be considered, just that (in my books) it's not a bug but an enhancement request, and hence it won't happen this late in the beta phase.

**Comment by andreas-pgss** [comment]
Thank you for your clarification, Eric,

May be the simplest solution would be a switch (command line parameter) to switch off the functionality that checks the end part of a path for being a symlink, something like "--prevent-symlink" with boolean values, true being the default so the current behaviour doesn't change by default.

**Comment by ericzolf** [comment]
I'm not so much concerned about this aspect. My concern is due to my experience that such last minute changes are often the ones which create trouble just before you release. Your issue is anyway part of 2.4.0 milestone for now.

In the meantime, perhaps have a look at bind-mounts: `mount -o bind /source/dir /target/dir` allows you to mount any directory on top of any other directory. I used it successfully for an application which, also, didn't like symlinks. There are  at least two drawbacks:

1. you need to be root to create bind-mounts (or at least have the necessary sudo rights)
2. relative symlinks pointing outside of the directory might point to the wrong place (but rdiff-backup doesn't care much, it just saves the symlinks as-is)

This said, I'm still not sure what you're exactly trying to reach because if you have access to the symlink, and using the symlink, you expect rdiff-backup to have access to the actual directory, then you have access to the actual directory, and can directly point rdiff-backup to it. As rdiff-backup stores files only relatively to the base directory, the result would be exactly the same.

In other words, _if_ symlinks would be supported, and you'd do `ln -s  /source/dir /target/dir`:

1. Calling `rdiff-backup /source/dir /mybackup` would give you exactly the same result as `rdiff-backup /target/dir /mybackup`, absolutly no difference (same applies to bind-mounts BTW).
2. If rdiff-backup can't reach `/target/dir`, then a symlink or a bind mount won't make it more possible to reach it and back-up what's in there

That's unless I've really misunderstood your issue or how ISPConfig can possibly work.

**Comment by andreas-pgss** [comment]
> I'm not so much concerned about this aspect. My concern is due to my experience that such last minute changes are often the ones which create trouble just before you release. Your issue is anyway part of 2.4.0 milestone for now.

That is fine with me. Thanks.

> In the meantime, perhaps have a look at bind-mounts: `mount -o bind /source/dir /target/dir`  [...]

Will check this out.

> This said, I'm still not sure what you're exactly trying to reach because if you have access to the symlink, and using the symlink, you expect rdiff-backup to have access to the actual directory, then you have access to the actual directory, and can directly point rdiff-backup to it. As rdiff-backup stores files only relatively to the base directory, the result would be exactly the same.

The problem is not backing up a single vhost, but backing up _n_ vhosts individually using a script with individual config files per vhost that use paths of the kind /var/www/fully.quallified.domain.name . I would need to completely rewrite the config files and the script to make it somehow use the "real" paths to the directories.

> That's unless I've really misunderstood your issue or how ISPConfig can possibly work.

The thing is: ISPConfig creates folders per customer and per vhost, so you get
/var/www/clients/client<m>/web<n>
But ISPConfig also creates the FQDN symlinks /var/www/fully.quallified.domain.name that are a lot easier to use for many purposes. This is the first time I find this to be a problem.

Thank you for taking the time to suggest a workaround and looking for other options.

**Comment by ericzolf** [comment]
Understood better your issue I think: you could call rdiff-backup on `$(realpath /var/www/fully.quallified.domain.name)` or `$(readlink -n -e /var/www/fully.quallified.domain.name)`.

**Comment by andreas-pgss** [comment]
> Understood better your issue I think: you could call rdiff-backup on `"$(realpath /var/www/fully.quallified.domain.name)"` or `"$(readlink -n -e /var/www/fully.quallified.domain.name)"`.

Thank you. I accidentally stumbled upon readlink myself yesterday and saw it as an option. But will check out realpath, too.

Cheers

Andreas

**Comment by andreas-pgss** [comment]
> But will check out realpath, too.

Actually, realpath seems to be the better option as it always returns the real path while readlink will only print the path if it ends in a link, otherwise nothing (so requires something like `[[ -L "$is_symlink_or_not" ]] && some_action`.
